google-cloud-clients/google-cloud-contrib/google-cloud-spanner-jdbc/src/main/java/com/google/cloud/spanner/jdbc/CredentialsService.java

```diff
@@ -0,0 +1,163 @@
+        return internalGetApplicationDefault();
+      } else {
+        return getCredentialsFromUrl(credentialsUrl);
+      }
+    } catch (IOException e) {
+      throw SpannerExceptionFactory.newSpannerException(
+          ErrorCode.INVALID_ARGUMENT, "Invalid credentials path specified", e);
+    }
+  }
+
+  @VisibleForTesting
+  GoogleCredentials internalGetApplicationDefault() throws IOException {
+    return GoogleCredentials.getApplicationDefault();
+  }
+
+  private GoogleCredentials getCredentialsFromUrl(String credentialsUrl) throws IOException {
+    Preconditions.checkNotNull(credentialsUrl);
+    Preconditions.checkArgument(
+        credentialsUrl.length() > 0, "credentialsUrl may not be an empty string");
+    if (credentialsUrl.startsWith(GOOGLE_CLOUD_STORAGE_PREFIX)) {
+      return getCredentialsFromCloudStorage(credentialsUrl);
+    } else {
+      return getCredentialsFromLocalFile(credentialsUrl);
+    }
+  }
+
+  private GoogleCredentials getCredentialsFromLocalFile(String filePath) throws IOException {
+    File credentialsFile = new File(filePath);
+    if (!credentialsFile.isFile()) {
+      throw new IOException(
+          String.format("Error reading credential file %s: File does not exist", filePath));
+    }
+    try (InputStream credentialsStream = new FileInputStream(credentialsFile)) {
+      return GoogleCredentials.fromStream(
+          credentialsStream, CloudSpannerOAuthUtil.HTTP_TRANSPORT_FACTORY);
+    }
+  }
+
+  private GoogleCredentials getCredentialsFromCloudStorage(String credentialsUrl)
+      throws IOException {
+    Preconditions.checkArgument(credentialsUrl.startsWith(GOOGLE_CLOUD_STORAGE_PREFIX));
+    try {
+      Storage storage = internalCreateStorage();
+      String bucketName = internalGetBucket(credentialsUrl);
+      String blobName = internalGetBlob(credentialsUrl);
+      Blob blob = storage.get(bucketName, blobName);
+      return GoogleCredentials.fromStream(
+          internalCreateInputStream(blob), CloudSpannerOAuthUtil.HTTP_TRANSPORT_FACTORY);
+    } catch (Exception e) {
+      throw new IOException(
+          String.format(
+              "Could not retrieve credentials from %s. "
+                  + "Credentials on Google Cloud Storage can only be used when the application is running "
+                  + "in an environment that already has default credentials to access Google Cloud Storage",
+              credentialsUrl),
+          e);
+    }
+  }
+
+  @VisibleForTesting
+  Storage internalCreateStorage() {
+    return StorageOptions.newBuilder().build().getService();
+  }
+
+  @VisibleForTesting
+  InputStream internalCreateInputStream(Blob blob) {
+    return Channels.newInputStream(blob.reader());
+  }
+
+  @VisibleForTesting
+  String internalGetBucket(String storageUrl) {
+    Preconditions.checkArgument(
+        storageUrl.startsWith(GOOGLE_CLOUD_STORAGE_PREFIX),
+        String.format("Storage URL must start with %s", GOOGLE_CLOUD_STORAGE_PREFIX));
```

Comment:
nit: not sure how often this gets called (likely just once) but we could make this a constant for use here and in `internalGetBlob`

Reply:
Moved to a constant.